### PR TITLE
Add an automatically generated Ignition provisioning token

### PR DESCRIPTION
**Pull request by cgwalters**
Part of implementing: https://github.com/openshift/enhancements/pull/443

The installer generates a random token (~password) and injects
it into the Ignition pointer configuration and as a secret into the
cluster.

The MCO will check it.

### CODE REVIEW

**Comment by cgwalters** [comment]
Hmm...not sure what's up with that e2e-aws run.  I can't find a must-gather or extracted cluster logs in the artifacts?
/test e2e-aws

File: pkg/asset/cluster/cluster.go [hunk 1]
```diff
@@ -51,6 +52,7 @@ func (c *Cluster) Dependencies() []asset.Asset {
 		&quota.PlatformQuotaCheck{},
 		&TerraformVariables{},
 		&password.KubeadminPassword{},
+		&ignition.ProvisioningToken{},
```
Comment (staebler) [inline pkg/asset/cluster/cluster.go:55]: Why does the Cluster asset need this dependency?

File: pkg/asset/ignition/machine/node.go [hunk 2]
```diff
@@ -37,16 +37,19 @@ func pointerIgnitionConfig(installConfig *types.InstallConfig, rootCA []byte, ro
 			ignitionHost = net.JoinHostPort(installConfig.VSphere.APIVIP, "22623")
 		}
 	}
+	queryvals := url.Values{}
+	queryvals.Add("token", token)
```
Comment (staebler) [inline pkg/asset/ignition/machine/node.go:41]: Move this inside the function where the URL is created.

File: pkg/asset/ignition/provisioningtoken.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+// Generate the token
+func (a *ProvisioningToken) Generate(asset.Parents) error {
+	b := make([]byte, tokenLen)
+	n, err := rand.Read(b)
+	if err != nil {
+		return err
+	}
+	if n != tokenLen {
```
Comment (staebler) [inline pkg/asset/ignition/provisioningtoken.go:36]: This check is not necessary. It is already guaranteed by the `rand.Read` function.

File: pkg/asset/ignition/machine/node.go (hunk 2 above)
Comment (cgwalters) [reply pkg/asset/ignition/machine/node.go:41]: This is that function, right?

File: pkg/asset/ignition/provisioningtoken.go (hunk 3 above)
Comment (cgwalters) [reply pkg/asset/ignition/provisioningtoken.go]: Yeah, excess of paranoia there.  Removed.

File: pkg/asset/ignition/machine/node.go (hunk 2 above)
Comment (cgwalters) [reply pkg/asset/ignition/machine/node.go]: Oh sorry misread.  Fixed.

**Comment by cgwalters** [comment]
Hmm I'm confused the contexts say "Job failed" yet it looks like e.g. the gofmt and unit contexts succeeded?

**Review by staebler — COMMENTED** [review]
/lgtm

/hold
on approval of the enhancement

**Pull request closed without merging**